pkg/kubelet/cpumanager/util.go

```diff
@@ -25,114 +25,147 @@ import (
-	result := cpuset.NewCPUSet()
-
-	if numCPUs == 0 {
-		return result, nil
-	}
-	if numCPUs > availableCPUs.Size() {
-		return nil, fmt.Errorf("not enough cpus available to satisfy request")
-	}
+type cpuAccumulator struct {
+	topo          *topology.CPUTopology
+	topoDetails   topology.CPUDetails
+	availableCPUs cpuset.CPUSet
+	numCPUsNeeded int
+	result        cpuset.CPUSet
+}
 
-	// Algorithm: topology-aware best-fit
-	CPUsPerCore := topo.NumCPUs / topo.NumCores
-	CPUsPerSocket := topo.NumCPUs / topo.NumSockets
-	topoDetails := topo.CPUtopoDetails.KeepOnly(availableCPUs)
-
-	// Auxilliary closure to update intermediate results:
-	// - Adds cpus to result
-	// - Recalculates availableCPUs
-	// - Prunes topoDetails
-	// - decrements numCPUs
-	take := func(cpus cpuset.CPUSet) {
-		result = result.Union(cpus)
-		availableCPUs = availableCPUs.Difference(result)
-		topoDetails = topoDetails.KeepOnly(availableCPUs)
-		numCPUs -= cpus.Size()
+func newCPUAccumulator(topo *topology.CPUTopology, availableCPUs cpuset.CPUSet, numCPUs int) *cpuAccumulator {
+	return &cpuAccumulator{
+		topo:          topo,
+		topoDetails:   topo.CPUtopoDetails.KeepOnly(availableCPUs),
+		availableCPUs: availableCPUs,
+		numCPUsNeeded: numCPUs,
+		result:        cpuset.NewCPUSet(),
 	}
+}
 
-	// Returns true if the supplied socket is fully available in
-	// `topoDetails`.
-	isFullSocket := func(socketID int) bool {
-		return topoDetails.CPUsInSocket(socketID).Size() == CPUsPerSocket
-	}
+func (a *cpuAccumulator) take(cpus cpuset.CPUSet) {
+	a.result = a.result.Union(cpus)
+	a.availableCPUs = a.availableCPUs.Difference(a.result)
+	a.topoDetails = a.topoDetails.KeepOnly(a.availableCPUs)
+	a.numCPUsNeeded -= cpus.Size()
+}
 
-	// Returns true if the supplied core is fully available in
-	// `topoDetails`.
-	isFullCore := func(coreID int) bool {
-		return topoDetails.CPUsInCore(coreID).Size() == CPUsPerCore
-	}
+// Returns true if the supplied socket is fully available in `topoDetails`.
+func (a *cpuAccumulator) isSocketFree(socketID int) bool {
+	return a.topoDetails.CPUsInSocket(socketID).Size() == a.topo.CPUsPerSocket()
+}
 
-	// 1. Acquire whole sockets, if available and the container requires
-	//    at least a socket's-worth of CPUs.
-	for _, s := range topoDetails.Sockets().Filter(isFullSocket).AsSlice() {
-		if numCPUs >= CPUsPerSocket {
-			glog.V(4).Infof("[cpumanager] takeByTopology: claiming socket [%d]", s)
-			take(topoDetails.CPUsInSocket(s))
-			if numCPUs < 1 {
-				return result, nil
-			}
-		}
-	}
+// Returns true if the supplied core is fully available in `topoDetails`.
+func (a *cpuAccumulator) isCoreFree(coreID int) bool {
+	return a.topoDetails.CPUsInCore(coreID).Size() == a.topo.CPUsPerCore()
+}
 
-	// 2. Acquire whole cores, if available and the container requires
-	//    at least a core's-worth of CPUs.
+// Returns socket IDs as a slice sorted by the number of whole available
+// cores, ascending
```

Comment:
> slice sorted by the number

The only sorting here is in `AsSlice()` which sorts sockets in ascending order(by socket ID), cant see any sorting based 

> whole available cores

Missing sort from previous implementation ? 
https://github.com/ConnorDoyle/kubernetes/blob/8cf339f7842e719d0202e3bbb48c90f72ea01a25/pkg/kubelet/cpumanager/util.go#L85